### PR TITLE
Global Styles: allow read access to users with `edit_posts` capabilities

### LINKED ISSUE #64755: Block Style Variations CSS not loaded for non-admins
### Description

The styles for block style variations are not visible in the editor for anyone without the `edit_theme_options` capability, so anyone but administrators.

They are loaded in the frontend but not in the backend.

Looking at the browser console one can see errors wit the REST API, but only in WP 6.6.1 _without_ Gutenberg. With Gutenberg I don't see those errors, but the behavior and display of the blocks in the editor still has the same bug.

### Step-by-step reproduction instructions

1. Create a user with any role that doesn't have `edit_theme_options` , e.g. `editor`.
2. Open a post and insert a button with a block style variation, e.g. `core/button`.
3. Apply a block style variation. You will **not** see the corresponding styles.
4. Save and open the post in the frontend and see the style.

### Screenshots, screen recording, code snippet

Screenshot showing 2 button blocks in the backend, looking the same although one has the outline style applied

Screenshot showing the same post in the frontend where everything is right

[a Video showing the whole flow with the difference between admin and editor](https://github.com/user-attachments/assets/d2f1483c-6c1e-4f19-a39a-b2adb8925e1f)

REST API errors in WP 6.6.1 without Gutenberg

Bildschirmfoto vom 2024-08-23 18-13-51

### Environment info

- WordPress 6.6.1
- with and without Gutenberg 19.0.0
- Twenty Twenty-Four theme

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by ramonjd**
Builds upon the base idea in:

- https://github.com/WordPress/gutenberg/pull/64797

With a view to:

- resolve https://github.com/WordPress/gutenberg/issues/64755
- make baby steps towards https://github.com/WordPress/gutenberg/issues/64670

## What

Allows any role that can edit a post, including custom post types, to read global styles from the API and in the editor.

## Why

Block style variations and other global styles are not available to non-admin editors. See https://github.com/WordPress/gutenberg/issues/64755

Furthermore, having global styles available in the post editor will, one day, allow block controls to reflect any values inherited from the theme/global styles. See: https://github.com/WordPress/gutenberg/issues/64670

### How

- Updating `read` capabilities to `edit_posts` in the global styles CPT (via filter)
- In the global styles API, updating permissions checks for GET requests to test for `current_user_can( 'edit_posts' )`
- In the global styles provider, remove any restrictions surrounding reading base config styles, and doing a "can update" check on the user global styles to either return the global styles editedEntity (for admins) or entity.

## Testing Instructions
1. As admin, make some changes to global styles in the site editor and save. This ensure there's a global styles CPT available.
2. Switch to an editor or author role.
3. Create a new post.
4. Run `await wp.data.resolveSelect( 'core/block-editor').getSettings()` in the browser console.
5. In the response, check the value of `Symbol(globalStylesDataKey)` - it should be a merged object of base theme and user global styles.
6. To check that you can't edit the global styles CPT, run `await wp.data.resolveSelect( 'core' ).getEntityRecord( 'root', 'globalStyles', await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentGlobalStylesId(), { context: 'edit' } )` in the console. You should see a 403.
7. Now switch to your admin profile.
8. Edit global styles in the Site Editor. Now reset or roll back using global styles revisions.  Make sure it all works as expected. 

Run a test or two:

```
npm run test:unit:php:base -- --filter WP_REST_Global_Styles_Controller_Gutenberg_Test
```

### CODE REVIEW

**Comment by Mamaduka** [comment]
We should also re-evaluate the `read` permissions check on the client side for global styles. They were introduced to avoid request failure for low-capability users but shouldn't be necessary with `read_posts` capability.

**Comment by ramonjd** [comment]
> We should also re-evaluate the read permissions check on the client side for global styles

Thanks for checking this out @Mamaduka 🙇🏻 

Could you explain this a bit more to me? Do you mean we probably don't need to do a `canUser( 'read' )` check on the `globalStyles` entity at all?

**Comment by Mamaduka** [comment]
> Do you mean we probably don't need to do a canUser( 'read' ) check on the globalStyles entity at all?

Correct. IIRC, anyone who can access editors will have this permission, but let's double-check when the REST controller update is ready. Happy to help with testing.

File: packages/editor/src/components/global-styles-provider/index.js
Comment (ramonjd) [inline packages/editor/src/components/global-styles-provider/index.js]: We could probably remove this since all roles that can read a post in the editor will need to read access to the styles object.

File: packages/editor/src/components/global-styles-provider/index.js
Comment (ramonjd) [inline packages/editor/src/components/global-styles-provider/index.js]: Only admins will be able view the "edited" entity, when update styles in the site editor

In the post editor, it's enough to grab the saved entity.

File: lib/compat/wordpress-6.7/rest-api.php
Comment (ramonjd) [inline lib/compat/wordpress-6.7/rest-api.php]: Because read access permissions are looser, we don't need the `edit` response to be preloaded. The view/read is enough.

**Review by ramonjd — COMMENTED** [review]
Added a test

File: packages/editor/src/components/global-styles-provider/index.js
Comment (aaronrobertshaw) [reply packages/editor/src/components/global-styles-provider/index.js]: This makes sense to me 👍 

_Edit_: I guess we could consider it a further optimization after https://github.com/WordPress/gutenberg/pull/65177. What's faster than not having to perform the check at all 😁

Comment (Mamaduka) [reply packages/editor/src/components/global-styles-provider/index.js]: > What's faster than not having to perform the check at all 😁

That's the most performant "check" 😄

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (aaronrobertshaw) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php]: Nit: Should the slug here reflect that this check is about read access to global styles not managing them?

File: lib/compat/wordpress-6.7/rest-api.php
Comment (aaronrobertshaw) [inline lib/compat/wordpress-6.7/rest-api.php:127]: Nit: Should this doc block also contain `@return`?

**Review by aaronrobertshaw — COMMENTED** [review]
This is testing well for me @ramonjd 👍 

For admin users:
✅ Global Styles resets, theme style variation switches, style updates, push instance styles globally etc all work as before.
✅ Within the post editor the admin user still gets the Global Styles data.

For non-admin users:
✅ Global Styles data is now available and things such as block style variations display correctly, resolving the issue highlighted by https://github.com/WordPress/gutenberg/issues/64755.

I think this is ready for some broader input from the REST API experts. Nice work!

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [reply lib/class-wp-rest-global-styles-controller-gutenberg.php]: Definitely!! I was hiding all the refinements behind the "Draft" status :trollface:

File: lib/compat/wordpress-6.7/rest-api.php
Comment (Mamaduka) [inline lib/compat/wordpress-6.7/rest-api.php]:
```suggestion
 * @return array Array of arguments for registering a post type.
```

This filter shouldn't return a `string`.

Comment (ramonjd) [reply lib/compat/wordpress-6.7/rest-api.php]: Will do. I did a straight copy paste from the [consuming method](https://github.com/WordPress/wordpress-develop/blob/47122105f20624e5a6b8ed151d732157f2f2605c/src/wp-includes/class-wp-post-type.php#L477)

**Comment by Mamaduka** [comment]
What's the reason for having the plugin override class directly in the `lib` directory instead of the specific `compat` folder?

The `WP_REST_Global_Styles_Controller` has been in the core since WP 5.9, so technically, I think we should extend it to avoid duplication.

File: packages/editor/src/components/global-styles-provider/index.js
Comment (Mamaduka) [inline packages/editor/src/components/global-styles-provider/index.js:80]: Looking at the [item schema definition](https://github.com/WordPress/gutenberg/blob/f073eb1f9f3b900ca347c7c8132fc7646f1bfaea/lib/class-wp-rest-global-styles-controller-gutenberg.php#L532-L573), I see that every property is exposed in `view` context.

I've not tested it, but this conditional call to selectors might not be needed. We should be able to call `getEditedEntityRecord` with `content: view` unless we're planning to review the schema and context for each property.

Comment (ramonjd) [reply packages/editor/src/components/global-styles-provider/index.js:80]: Isn't `edit` the default context in the editor thanks to the entity base URL params?

https://github.com/WordPress/gutenberg/blob/bdfd4696797f19f844057b88664dd48ba2155de7/packages/core-data/src/entities.js#L185-L185

If not, great! 🎉 

Otherwise, I did it the "longhand" way with "view" for now to avoid having to backtracking and adding `edit` in the appropriate places. Was planning on doing that in a follow up.

**Comment by ramonjd** [comment]
Thanks for testing again, folks! 🍺 

> What's the reason for having the plugin override class directly in the lib directory instead of the specific compat folder?

For similar reasons to when the inheritance chain was canned and the full `WP_Theme_JSON_Gutenberg` class was bundled in https://github.com/WordPress/gutenberg/pull/46579

The REST controller followed suit based on the same experience: https://github.com/WordPress/gutenberg/pull/53618

The smallest inheritance override sometimes entailed copying half the class due to the prevalence of `_Gutenberg` classes.

File: packages/editor/src/components/global-styles-provider/index.js
Comment (Mamaduka) [reply packages/editor/src/components/global-styles-provider/index.js:80]: Yes, `edit` is the default context.

I forgot that `getEditedEntityRecord` doesn't accept the `query` parameter :/

**Review by spacedmonkey — DISMISSED** [review]
I am not sure about this change. This could be seen as a breaking change. Not all users should have access to this data. I am against this change. Thoughts @TimothyBJacobs

**Comment by talldan** [comment]
@spacedmonkey I don't think it's good behavior to use the 'request changes' feature on Github without actually requesting any changes. It leaves the PR in limbo.

It's fine if you have concerns, but @ramonjd isn't going to merge the PR on you without discussing them first, and even then it can be reverted.

It'd be great if you could expand upon your comments.

Also worth noting that @TimothyBJacobs was involved in the conversation around this change earlier - https://github.com/WordPress/gutenberg/pull/64797#discussion_r1742093233.

**Comment by ramonjd** [comment]
> I am not sure about this change. This could be seen as a breaking change. Not all users should have access to this data. I am against this change.

I'm happy to learn there **might** be a better way to do things, but it would be nice to hear what you think that would be or where the problem lies exactly. 

1. It is specifically the changes to the post type caps, the endpoint, or what?
2. Where in the codebase should contributors be looking exactly to progress to this issue?

Posting the original trac issue for context: https://core.trac.wordpress.org/ticket/54516

Let's get more feedback from folks to find out the most optimal way forward.

**Comment by peterwilsoncc** [comment]
I left code notes on the WP-Dev PR which Ramon is clever enough to figure out apply here too. 

> Not all users should have access to this data. I am against this change.

For the current theme, the data is public in it's rendered form via the style element added in the HTML header so I don't think allowing increased access via the REST API is overly concerning in terms of exposed data.

With the limited approach this PR takes, I think it can be viewed as a bug fix for #64755.

**Comment by aaronrobertshaw** [comment]
> For the current theme, the data is public in it's rendered form via the style element added in the HTML header so I don't think allowing increased access via the REST API is overly concerning in terms of exposed data.

Agreed. 

> With the limited approach this PR takes, I think it can be viewed as a bug fix for https://github.com/WordPress/gutenberg/issues/64755.

+1

I might be missing something but I don't think the proposed approach prevents us from iterating on these permissions.

**Comment by spacedmonkey** [comment]
Can we wait until https://github.com/WordPress/gutenberg/pull/65071 is merge before merging this change. It will effect this PR in lots of ways.

**Comment by spacedmonkey** [comment]
Let's not merge this until https://github.com/WordPress/gutenberg/pull/65259 is merged.

**Comment by ramonjd** [comment]
> Let's not merge this until https://github.com/WordPress/gutenberg/pull/65259 is merged.

Yep, already on the radar. 👍🏻 Thanks.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for iterating here and on https://github.com/WordPress/gutenberg/pull/65259 @ramonjd 👍 

I've given this another run and it is testing pretty well for me. I did notice things have regressed a little compared to trunk with the preloading of global styles paths. I realise it isn't perfect on trunk either but it would be good to not get further from the ideal.

Getting the preloaded global styles data sorted could be a follow up if you think it's best to get this bug fix into the next GB release.

Here's what I'm seeing on this branch:

| Admin User | Non-admin User|
|---|---|
| <video src="https://github.com/user-attachments/assets/e4a15051-f289-4cc8-b62e-e3bb0e019be0" /> | <video src="https://github.com/user-attachments/assets/c890b531-f061-497e-920b-4f2c5d53f0b9" /> |


Here's what I get on trunk for an admin user (non-admin has the bug this PR addresses):


https://github.com/user-attachments/assets/c2abbf97-15f6-4f8b-849f-b96d2ea825ea

**Comment by ramonjd** [comment]
>  I did notice things have regressed a little compared to trunk with the preloading of global styles paths. I realise it isn't perfect on trunk either but it would be good to not get further from the ideal.

Agreed. Thanks for testing that. I'll take a look. It would be good not to introduce a regression.

File: lib/compat/wordpress-6.7/rest-api.php
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.7/rest-api.php]: Quickly hacking around to see the impact of these changes in relation to the [regression noted earlier](https://github.com/WordPress/gutenberg/pull/65071#pullrequestreview-2308168250), removing this change brings the flash of styling much closer to trunk for both admin and non-admin users. Simply including the view context path as well didn't appear to make much difference.

Comment (ramonjd) [reply lib/compat/wordpress-6.7/rest-api.php]: I can reproduce. I'll remove this change 👍🏻 and test the preloading array in a follow up. Thanks for finding it.

Logically, all we need is the view preloaded, but I suspect there's something going on at a deeper level that's evading me.

Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.7/rest-api.php]: Yeah, I'm missing something here as well. Seems related to the fact we get either the edit or view context based on user caps. I was wondering if the preloaded paths needed the same conditional treatment.

That said, I'll need to re-test to see if the non-admin user faced the same issue before the latest commit.

**Comment by ramonjd** [comment]
While testing with @aaronrobertshaw we noticed that was no difference between preloading global styles for the post editor and not doing it, which means that something is buggy. Or maybe not 😄 

I removed the update to the preload filter path array for now.

I think we should look at this in an immediate follow up as it might be an unrelated bug.

**Review by aaronrobertshaw — APPROVED** [review]
> While testing with @aaronrobertshaw we noticed that was no difference between preloading global styles for the post editor and not doing it, which means that something is buggy. Or maybe not 😄

While giving this PR a final test in its current state, I think preloading the view context for global styles does have a tiny reduction in the flash of styling for the non-admin user in the post editor. 

This isn't enough to overcome the real issue of a styling flash though. More on that below.

> I think we should look at this in an immediate follow up as it might be unrelated bug.

Agreed. 

As it stands, this PR solves the main bug and the styling flash is the same as on trunk for admin users. A dedicated follow-up specifically tackling the preloading of global styles data sounds like a good idea. 

With all that said, this PR is testing as advertised.

For admin users:
✅ Global Styles resets, theme style variation switches, style updates, push instance styles globally etc all work as before.
✅ Within the post editor the admin user still gets the Global Styles data.

For non-admin users:
✅ Global Styles data is now available and things such as block style variations display correctly, resolving the issue highlighted by https://github.com/WordPress/gutenberg/issues/64755.

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (spacedmonkey) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:562]: Why do the check for edit theme files removed here? There maybe cases where user might not have edit post but might have edit theme files. 

We can keep the existing check as before.

**Review by spacedmonkey — DISMISSED** [review]

File: lib/class-wp-rest-global-styles-controller-gutenberg.php
Comment (ramonjd) [reply lib/class-wp-rest-global-styles-controller-gutenberg.php:562]: Done, with tests.

**Review by aaronrobertshaw — APPROVED** [review]
The latest changes make sense and address the concern raised in https://github.com/WordPress/gutenberg/pull/65071#discussion_r1764116237. Personally, I think that scenario was unlikely and overkill but better safe than sorry 👍 

✅ Tests are passing locally
✅ Smoke test didn't reveal any regressions

LGTM 🚢

**Comment by ramonjd** [comment]
Thank you! I'll update the Core backport PR so discussion can move there.

Thanks for the feedback, everyone.

**Comment by fabiankaegy** [comment]
Hey @ramonjd   👋 

Would you be able to help write a dev note for this for the 6.7 release? We are planning to have this as part of a larger Miscellaneous Editor Updates note.

We are hoping to get all drafts in by October 13th to leave some time for reviews before the RC1. 

All Dev Notes get tracked in #65784 so feel free to leave a note there or ping me directly :) 

Please let us know if you can assist with that.

Thanks in advance :)

**Comment by aaronrobertshaw** [comment]
> We are planning to have this as part of a larger Miscellaneous Editor Updates note.

The scope of this change is pretty narrow, so being under the banner of Miscellaneous Editor Updates sounds good 👍 

To help get the ball rolling I've drafted something small below. I'll leave it for Ramon to iterate on (or discard if its off the mark).

---

### Dev Note: Read access to Global Styles data

In WordPress 6.7, users with the `edit_posts` capability will [gain permission](https://github.com/WordPress/gutenberg/pull/65071) to access raw Global Styles data. This change supports recent improvements to [block style variations](https://make.wordpress.org/core/2024/06/24/section-styles/) and upcoming UI enhancements in [displaying style inheritance](https://github.com/WordPress/gutenberg/issues/37752).

**Pull request merged**
